codec-http/src/test/java/io/netty/handler/codec/http/HttpServerUpgradeHandlerTest.java

```diff
@@ -0,0 +1,124 @@
+import io.netty.util.CharsetUtil;
+
+import static org.junit.Assert.*;
+
+public class HttpServerUpgradeHandlerTest {
+
+    private class TestUpgradeCodec implements UpgradeCodec {
+        @Override
+        public Collection<CharSequence> requiredUpgradeHeaders() {
+            return ImmutableList.of();
+        }
+
+        @Override
+        public boolean prepareUpgradeResponse(ChannelHandlerContext ctx, FullHttpRequest upgradeRequest,
+                                              HttpHeaders upgradeHeaders) {
+            return true;
+        }
+
+        @Override
+        public void upgradeTo(ChannelHandlerContext ctx, FullHttpRequest upgradeRequest) {
+            // Ensure that the HttpServerUpgradeHandler is still installed when this is called
+            assertEquals(ctx.pipeline().context(HttpServerUpgradeHandler.class), ctx);
+            assertNotNull(ctx.pipeline().get(HttpServerUpgradeHandler.class));
+
+            // Add a marker handler to signal that the upgrade has happened
+            ctx.pipeline().addAfter(ctx.name(), "marker", new ChannelInboundHandlerAdapter());
+          }
+    }
+
+    @Test
+    public void upgradesPipelineInSameMethodInvocation() {
+        final HttpServerCodec httpServerCodec = new HttpServerCodec();
+        final UpgradeCodecFactory factory = new UpgradeCodecFactory() {
+            @Override
+            public UpgradeCodec newUpgradeCodec(CharSequence protocol) {
+                return new TestUpgradeCodec();
+            }
+        };
+
+        ChannelHandler testInStackFrame = new ChannelDuplexHandler() {
+            // marker boolean to signal that we're in the `channelRead` method
+            private boolean inReadCall;
+            private boolean wroteMessage;
+
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                assertFalse(inReadCall);
+                assertFalse(wroteMessage);
+
+                inReadCall = true;
+                try {
+                    super.channelRead(ctx, msg);
+                    // All in the same call stack, the upgrade codec should receive the message,
+                    // written the upgrade response, and upgraded the pipeline.
+                    assertTrue(wroteMessage);
+                    assertNull(ctx.pipeline().get(HttpServerCodec.class));
+                    assertNotNull(ctx.pipeline().get("marker"));
+                } finally {
+                    inReadCall = false;
+                }
+            }
+
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                // We ensure that we're in the read call, but we don't flush the message
+                // to make sure that the pipeline upgrade machinery isn't dependant on
+                // the message being flushed.
+                assertTrue(inReadCall);
```

Comment:
Call `super.write(....)` to ensure we not "leak the message" and also to ensure the promise is notified.

Reply:
This was deliberate so that we could be confident that the pipeline reshaping happened in this call stack irrespective of the promise being satisfied.

That said, perhaps it's not something that is likely to regress. And we do need to do something with the outbound message, good catch.